hibernate-envers/src/main/java/org/hibernate/envers/query/internal/impl/RevisionsOfEntityAssociationQuery.java

```diff
@@ -0,0 +1,109 @@
+import org.hibernate.envers.internal.entities.EntityConfiguration;
+import org.hibernate.envers.internal.entities.mapper.id.IdMapper;
+import org.hibernate.envers.internal.reader.AuditReaderImplementor;
+import org.hibernate.envers.internal.tools.query.Parameters;
+import org.hibernate.envers.internal.tools.query.QueryBuilder;
+import org.hibernate.envers.query.AuditAssociationQuery;
+
+/**
+ * An {@link org.hibernate.envers.query.AuditAssociationQuery} implementation for {@link RevisionsOfEntityQuery}.
+ *
+ * @author Chris Cranford
+ */
+@Incubating
+public class RevisionsOfEntityAssociationQuery<Q extends AuditQueryImplementor> extends AbstractAuditAssociationQuery<Q> {
+
+    public RevisionsOfEntityAssociationQuery(
+            EnversService enversService,
+            AuditReaderImplementor auditReader,
+            Q parent,
+            QueryBuilder queryBuilder,
+            String propertyName,
+            JoinType joinType,
+            Map<String, String> aliasToEntityNameMap,
+            String ownerAlias,
+            String userSuppliedAlias) {
+        super( enversService,
+               auditReader,
+               parent,
+               queryBuilder,
+               propertyName,
+               joinType,
+               aliasToEntityNameMap,
+               ownerAlias,
+               userSuppliedAlias );
+    }
+
+    @Override
+    public AuditAssociationQuery<? extends AuditAssociationQuery<Q>> traverseRelation(
+            String associationName,
+            JoinType joinType,
+            String alias) {
+        AbstractAuditAssociationQuery<AbstractAuditAssociationQuery<Q>> result = associationQueryMap.get( associationName );
+        if ( result == null ) {
+            result = new RevisionsOfEntityAssociationQuery<>(
+                    enversService,
+                    auditReader,
+                    this,
+                    queryBuilder,
+                    associationName,
+                    joinType,
+                    aliasToEntityNameMap,
+                    this.alias,
+                    alias );
+
+            associationQueries.add( (RevisionsOfEntityAssociationQuery<Q>) result );
+            associationQueryMap.put( associationName, result );
+        }
+
+        return result;
+    }
+
+    @Override
+    protected void addCriterionsToQuery(AuditReaderImplementor versionsReader) {
+        if ( enversService.getEntitiesConfigurations().isVersioned( entityName ) ) {
+            String auditEntityName = enversService.getAuditEntitiesConfiguration().getAuditEntityName( entityName );
+            Parameters joinConditionParameters = queryBuilder.addJoin( joinType, auditEntityName, alias, false );
+
+            // owner.reference_id = target.originalId.id
+            AuditEntitiesConfiguration verEntCfg = enversService.getAuditEntitiesConfiguration();
+            final String prefix = alias.concat( "." ).concat( verEntCfg.getOriginalIdPropName() );
```

Comment:
seems 
```
final String prefix = alias + "." + verEntCfg.getOriginalIdPropName();
```
could be slightly better in terms of performance. see https://stackoverflow.com/questions/47605/string-concatenation-concat-vs-operator

not a big deal, though.

Reply:
Changed in multiple places, thanks for the suggestion.